site/app/libraries/database/DatabaseQueries.php

```diff
@@ -3651,6 +3651,12 @@ public function getCourseForUserId($user_id, bool $archived = false): array {
             $course->loadDisplayName();
             $return[] = $course;
         }
+
+        // Callback to filter out any courses a student has dropped so they do not appear on the homepage.
+        $return = array_filter($return, function (Course $course) use ($user_id) {
+            return $this->checkStudentActiveInCourse($user_id, $course->getTitle(), $course->getSemester());
```

Comment:
In moving this into the query function, we would also want to then just build it straight into the query itself, getting a bit more optimized in the number of queries we run on a page. In this case, we could add a `WHERE` clause of the effect `registration_section IS NOT NULL OR user_group<>4` and be equivalent to what this filter is doing. It should be fine to add this where clause in combination with the change the on the other file for the add course page.